policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,712 @@
+        let namespace = saz.namespace().unwrap();
+        let name = saz.name();
+        match server_authorization::ServerAuthz::from_resource(saz, &self.cluster_info) {
+            Ok(meta) => self
+                .namespaces
+                .get_or_default_with_reindex(namespace, move |ns| {
+                    ns.policy.update_server_authz(name, meta)
+                }),
+            Err(error) => tracing::error!(%error, "Illegal server authorization update"),
+        }
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        self.namespaces.get_with_reindex(namespace, |ns| {
+            ns.policy.server_authorizations.remove(&name).is_some()
+        })
+    }
+
+    fn reset(
+        &mut self,
+        sazs: Vec<k8s::policy::ServerAuthorization>,
+        deleted: HashMap<String, HashSet<String>>,
+    ) {
+        #[derive(Default)]
+        struct Ns {
+            added: Vec<(String, server_authorization::ServerAuthz)>,
+            removed: HashSet<String>,
+        }
+
+        // Aggregate all of the updates by namespace so that we only reindex
+        // once per namespace.
+        let mut updates_by_ns = HashMap::<String, Ns>::default();
+        for saz in sazs.into_iter() {
+            let namespace = saz
+                .namespace()
+                .expect("serverauthorization must be namespaced");
+            let name = saz.name();
+            match server_authorization::ServerAuthz::from_resource(saz, &self.cluster_info) {
+                Ok(meta) => updates_by_ns
+                    .entry(namespace)
+                    .or_default()
+                    .added
+                    .push((name, meta)),
+                Err(error) => tracing::error!(%error, "Illegal server authorization update"),
+            }
+        }
+        for (ns, names) in deleted.into_iter() {
+            updates_by_ns.entry(ns).or_default().removed = names;
+        }
+
+        for (namespace, Ns { added, removed }) in updates_by_ns.into_iter() {
+            if added.is_empty() {
+                self.namespaces.get_with_reindex(namespace, |ns| {
+                    let changed = !removed.is_empty();
+                    for name in removed.into_iter() {
+                        ns.policy.server_authorizations.remove(&name);
+                    }
+                    changed
+                });
+            } else {
+                self.namespaces
+                    .get_or_default_with_reindex(namespace, |ns| {
+                        let mut changed = !removed.is_empty();
+                        for name in removed.into_iter() {
+                            ns.policy.server_authorizations.remove(&name);
+                        }
+                        for (name, saz) in added.into_iter() {
+                            changed = ns.policy.update_server_authz(name, saz) || changed;
+                        }
+                        changed
+                    });
```

Comment:
the same comment on the reset impl for ``Server`s applies here as well, this if/else doesn't _look_ necessary to me?